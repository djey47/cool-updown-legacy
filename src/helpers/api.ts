/**
 * Provides utility functions to handle Cool-updown API requests and responses
 */

import { ApiRequest } from "../model/models";

/**
 * @param request API request from Express middleware
 * @returns server identifier which is passed through a path parameter
 */
export function extractServerIdentifier(request: ApiRequest) {
  return request.params.serverId;
}
