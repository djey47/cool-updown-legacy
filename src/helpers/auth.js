/* @flow */

const basicAuth = require('express-basic-auth');

/*
 * Generated by:
 * https://www.random.org/strings/?num=10&len=10&digits=on&upperalpha=on&loweralpha=on&unique=on&format=html&rnd=new
 */
const AUTH_REALM = 'NyDGMFeptk';

/**
 * Basic HTTP authentication
 */
function initBasicAuth(app, isEnabled, userName, password) {
  if (!isEnabled) return;

  app.use(basicAuth({
    users: { [userName]: password },
    challenge: true,
    realm: AUTH_REALM,
  }));
}

module.exports = {
  initBasicAuth,
};
