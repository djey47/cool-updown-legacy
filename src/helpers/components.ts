/*
 * Misc. page components
 */

/**
 * @returns contents in a pragraph then a link in a next one
 */
export function withBackLink(contents: string, linkUrl: string, linkText: string) {
    return `
    <p class="backlink-contents">${contents}</p>
    <p class="backlink-link"><a href="${linkUrl}">${linkText}</a></p>
    `;
}
