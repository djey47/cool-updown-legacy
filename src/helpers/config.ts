import config from 'config';
import { TypedResponse } from "../model/express";
import { ServerConfig } from '../model/models';
import messages from '../resources/messages';

/**
 * 
 * @param res 
 * @param serverId 
 * @returns 
 */
export function retrieveServerConfiguration(res: TypedResponse<string>, serverId: string) {
  const serversConfiguration = config.get('servers');
  const serverConfiguration = serversConfiguration[serverId] as ServerConfig;

  if (!serverConfiguration) {
    res.status(400).send(messages.errors.invalidArg);
  }

  return serverConfiguration;
}
