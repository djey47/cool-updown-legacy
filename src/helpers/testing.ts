/**
 * Unit testing support functions
 */

/**
 * @param expressResponseMock 
 * @returns an Express default response (with mocks) for unit tests
 */
export const generateDefaultResponse = (expressResponseMock) => ({
  status: c => expressResponseMock.statusMock(c),
  send: msg => expressResponseMock.sendMock(msg),
  json: ()  => expressResponseMock.jsonMock(),
});    
