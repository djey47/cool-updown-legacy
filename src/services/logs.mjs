import appRootDir from 'app-root-dir';
import { readFile } from 'fs/promises';
import path from 'path';
import logger from '../helpers/logger.mjs';

/**
 * Handles LOGS request
 */
export default async function logs(_req, res) {
  try {
    const logsContents = await readFile(path.join(appRootDir.get(), 'logs', 'app.log'));

    res.send(`<pre>${logsContents}<pre>`);
  } catch (error) {
    logger.error(`(logs) ${error}`);

    res.status(204).send();
  }
}
