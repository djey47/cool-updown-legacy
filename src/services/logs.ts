import appRootDir from 'app-root-dir';
import { readFile, stat } from 'fs/promises';
import path from 'path';
import logger from '../helpers/logger';
import { TypedResponse } from '../model/express';
import { generatePage } from '../helpers/page';
import { withBackLink } from '../helpers/components';
import messages from '../resources/messages';

/**
 * Handles LOGS request
 */
export default async function logs(_req: Express.Request, res: TypedResponse<string>) {
  try {
    const logFilePath = path.join(appRootDir.get(), 'logs', 'app.log');
    const fileStats = await stat(logFilePath);
    const logsContents = await readFile(logFilePath);
    const htmlResponse = `
    ${withBackLink(`${fileStats.size} ${messages.bytes}`, '/', messages.home)}
    ${withBackLink(`<pre>${logsContents}</pre>`, '/', messages.home)}
    `;
    res.send(generatePage(htmlResponse));
  } catch (error) {
    logger.error(`(logs) ${error}`);

    console.log(error);

    res.status(500).send(generatePage(withBackLink(messages.logs.none, '/', messages.home)));
  }
}
