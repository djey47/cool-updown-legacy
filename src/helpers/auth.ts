import basicAuth from 'express-basic-auth';
import config from 'config';
import { readFile } from 'fs/promises';

interface App {
  use: (plugin: unknown) => unknown;
}

/*
 * Generated by:
 * https://www.random.org/strings/?num=10&len=10&digits=on&upperalpha=on&loweralpha=on&unique=on&format=html&rnd=new
 */
const AUTH_REALM = 'NyDGMFeptk';

/**
 * Basic HTTP authentication
 */
export function initBasicAuth(app: App, isEnabled: boolean, userName: string, password: string) {
  if (!isEnabled) return;

  app.use(basicAuth({
    users: { [userName]: password },
    challenge: true,
    realm: AUTH_REALM,
  }));
}

/**
 * @private
 * @return Promise which resolves to String,
 * with the contents of SSH private key as per provided keyPath in configuration
 */
export async function readPrivateKey() {
  const privateKeyPath = config.get('server.keyPath') as string;
  const privateKey = await readFile(privateKeyPath, 'utf-8');

  // console.log('services::getPrivateKey', { privateKeyPath, privateKey });

  return privateKey as string;
}
