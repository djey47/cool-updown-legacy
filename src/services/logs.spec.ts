import globalMocks from '../../config/jest/globalMocks';
import resetMocks from '../../config/jest/resetMocks';
import { generateDefaultResponse } from '../helpers/testing';
import logs from './logs';

const {
  appRootDirMock, expressResponseMock, nodeFSMock,
} = globalMocks;

const res = generateDefaultResponse(expressResponseMock);

describe('logs service', () => {
  beforeEach(() => {
    resetMocks();

    // Async file reader OK with log contents
    nodeFSMock.readFile.mockResolvedValue('These are logs contents');

    expressResponseMock.statusMock.mockImplementation(() => expressResponseMock);
  });

  it('should return 204 if no log file', async () => {
    // given
    nodeFSMock.readFile.mockRejectedValue('file not found');
    appRootDirMock.get.mockImplementationOnce(() => '/foo/bar');

    // when
    await logs(undefined, res);

    // then
    expect(expressResponseMock.statusMock).toHaveBeenCalledWith(204);
    expect(expressResponseMock.sendMock).toHaveBeenCalledWith(undefined);
  });

  it('should generate correct response with log file', async () => {
    // given-when
    await logs(undefined, res);

    // then
    expect(expressResponseMock.sendMock).toHaveBeenCalled();
    expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
  });
});
