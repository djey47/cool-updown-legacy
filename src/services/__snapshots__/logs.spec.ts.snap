// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logs service should generate correct response with log file 1`] = `"<pre>These are logs contents<pre>"`;
